core/src/Kokkos_Complex.hpp

```diff
@@ -303,16 +316,45 @@ class alignas(2 * sizeof(RealType)) complex {
 
   //! Assignment operator, volatile LHS and volatile RHS
   // TODO Should this return void like the other volatile assignment operators?
+  //
+  // Templated, so as not to be a copy assignment operator (Kokkos issue #2577)
+  // Intended to behave as
+  //    volatile complex& operator=(const volatile complex&) volatile noexcept
+  //
+  // Use cases:
+  //    volatile complex vr;
+  //    const volatile complex cvr;
+  //    volatile complex vl;
+  //    vl = vr;
+  //    vl = cvr;
+  template <class Complex,
+            typename std::enable_if<std::is_same<Complex, complex>::value,
+                                    int>::type = 0>
   KOKKOS_INLINE_FUNCTION volatile complex& operator=(
```

Comment:
I think this has to return `void` IIUC.  There is some issue with returning references to `volatile` that I've come across several times.

Reply:
But it didn't before did it?

Reply:
At least some of those volatile overloads returned void.

Reply:
I tried not to change any of the return types from their original specification.  The only one that returned `void` (and still does) was:

`void operator=(const complex<RealType>& src) volatile noexcept { /* ... */ }`

Reply:
I didn't change any of the current signatures.

Reply:
There is a warning which is described in Issue #177.  Now that they are not copy assignment operators, it is easier to change their return type, but that was more than I was comfortable changing initially.